### DISCUSSION TITLE
What is the best practice to sort the behavior examples ?

### DISCUSSION THREAD

**Discussion by rombony** [umyelab/LabGym#119] [Q&A]
Hi, we wish to train a LabGym classifier to score scratching bouts in mice. We filmed black mice from the bottom on a glass plate. Pixel substraction didn't work well for us as paws and tail were often missed, probably due to poor contrast with the background. But training detectron2 on 150 annotated images seems to do the trick.

As we're now ready to score lots of examples, we're wondering what would be the best practice for labeling.

- When there is a transition between two behaviors in an example, should I label the behavior with the most frames or discard the example ? We thought to create a 'trash' behavior category to store those examples and delete it before to train the classifier.

- Do you have an idea of the best number of frames per example ? We used 25 frames in our tests.

- Would it be better to set a high number of behaviors to encompass all possibilities (but maybe the number of examples will be small in these categories), or to reduce the number of behaviors to the most common ones? Also, any advice on the best number of examples (per behavior or altogether) for training and testing the categorizer would be appreciated.

Many thanks for any tips you can give us, and congratulations for this amazing work!

Romane

**Comment by yujiahu415** [comment]
Hi Romane,

- You can sort that kind of examples as a new behavior category "xx transition", if that happens frequently. If it's rare, just discard it. Including a "trash" category is always a good idea and when train a classifier, INCLUDE that "trash" category as well. Basically, you need to sort all the behavior categories--the complete ethogram that show in your videos to analyze, including both the behaviors of interest and background behaviors that occurs frequently. For example, if behavior A and B are your interest, but C also show frequently in your videos to analyze, if you train a classifier only using A and B examples, when the classifier sees C, it will forcibly categorize it into A or B category, depending on the similarity between C and A or B, because the classifier doesn't know there's also a C.

- What is the fps of your videos? What's the duration of an episode of the behavior type that has the longest duration among all the behavior types? If the fps of your videos is 30, 25-frame is about 0.8 second, which seems to be a reasonable choice for common behaviors in rodent. We previously used 15-30 frames (0.5-1 second) at 30 fps and it worked well.

- If two behavior examples look very different to you, make them into two categories. If you cannot easily tell the difference between two examples by just watching the examples (animations and pattern images), then group them into one category. This is the general rule. The more distinct the examples from different categories look, the more accurate of the classifier. Typically, 50~200 pairs (a pair contains one animation and its paired pattern image) for each category are sufficient to train a classifier with good accuracy. You can sort as many categories as possible at first, and then revise them to see if two categories are similar to be grouped as one. If a behavior occurs rarely, just exclude it. For testing, the more testing examples, the higher statistical power, but you need at lease 10 pairs for each category to run the testing.

**Reply by rombony** [reply]
Many thanks for your help!

We’ll keep our trash category then, and include it in the training. We’ll use a ‘transition’ category as well (to be discarded before training).

Our videos are 50 fps. We generate examples of 25 frames, so 0.5 second.

We’ll make sure to use the whole ethogram. About that, would lateralized behaviors fall in the same category? By that I mean: is it better to distinguish between licking of the left and right hindpaw for example, and pool the results after the analysis if we’re not interested in lateralization?

**Reply by yujiahu415** [reply]
50 fps is quite a lot. There's a fps downsampling function in LabGym's Preprocessing Module for you to downsampling the fps, which will significantly increase the training and analyzing speed if the video duration is long. In our hands, 15~30 fps is sufficient for most rodent behaviors. But that's your decision. If you decide to downsampling fps, make sure to downsampling all the videos to the same fps because a Categorizer trained on a certain fps can only be useful for analyzing videos with the same fps.

If you don't care about distinguishing licking of the left hind paw vs licking of the right one, you can pool them into one category "licking hind paw", and including both left and right scenarios will increase the diversity of training examples for this category and will be beneficial to the training.

**Reply by rombony** [reply]
Thanks again for your help and your quick response, 

We will keep our videos at 50 fps because a scratching bout is a very fast movement and we haven't any constraints for the training duration. 

We also faced some issues with Detectron2, we train the detector with 150 annotated images leading the detector to a mean average precision of 89.2580%. We then generate behavior examples but some of them contain frames where the mouse disappears, do you have any idea where this could be coming from or if it is because the detector training is not sufficient enough ?

**Reply by yujiahu415** [reply]
150 annotated images should be good--if they cover most scenarios (you don't need to include images that look similar, e.g., the animal in the same place, doing the same thing).

When you generate dataset in roboflow, make sure to SKIP all the preprocessing steps (like resizing to 640), and make sure to select "flip, both horizontally and vertically" as the augmentation method. If there's illumination changes in your video, after generate a dataset with flip augmentation (which will give you ~3 folds of the original images), import all images to create a new project and choose "brightness, + - 10%" as augmentation method and generate a new dataset (~9 folds of original images). Train a new Detector on the "~9 folds" dataset. The accuracy should improve. When export the dataset, make sure to select "COCO instance segmentation" as format.

And also, what is the inferencing frame size when you train the Detector? And how many iterations for the training? What is the original frame size of your videos? And did you resize your videos when generating the examples?

**Reply by rombony** [reply]
We first tried training the detector with the following parameters : inferencing framesize at 1600 and iterations at 2000. This is were we had problems with mouse detection. Then we tried the default values and it solved the problem. 

We followed your advices and sorted 12k examples into 8 categories, ranging from 283 to 3041 examples. We did not include transitions, only clear behaviors. We did not include a trash either as there were only 6 occurences of the mouse jumping, which wasn't part of our ethogram.

We trained the classifier with the following parameters : 
-Complexity level 3/7 for both Animation Analyzer and Pattern Recognizer
-Input shape = 32 for both Animation Analyzer and Pattern Recognizer
-No Graysacle
-Number of frames for an animation / pattern image : 25
-With bodyparts
-STD at 0 
-With default augmentation and no augment validation data

We haven't sorted test examples yet, but looking at the videos we can tell that the behavior is not properly sorted. We noted that the training was done on very few iterations. It stopped at epoch 25/1 000 000 with a loss of loss: 0.2765.

From our previous deeplabcut experience, we were expecting much lower values. Is that normal ?

I've attached the log in case it's helpful.

[training log.txt](https://github.com/umyelab/LabGym/files/14704214/training.log.txt)

Again, thank you very much for your help!

**Reply by yujiahu415** [reply]
The training loss in LabGym is different from that in DLC. So they are not comparable. But 0.27 training loss (and 0.39 validation loss, which is a more decisive metric) is not good.

Several ways to try to improve the accuracy:

1. Since the amount of training examples in your dataset is way more than sufficient, you can refine the sorting by: 1) Remove redundant / increase diversity: if there are many examples that look the same in one category, just keep one pair. You only need to keep examples that cover different scenarios of the behavior. Make each category no more than 500 pairs of examples (one pair contains an animation and an image) 2) Make sure the examples in different categories look differently. If the examples in different behavior categories look similar, the training will not be good. Some behaviors may be difficult to distinguish from others. So you need to decide whether to combine them to the one category, or figure out a clear boundary that can distinguish examples in one category from those in another category.

2. When you watch the annotated videos, check whether the Detector works well to precisely detect the animal outline. If not, try to refine the Detector. And also, when you generated the behavior examples, did you use another Detector (with 1600 inferencing frame size) that performs worse? Are the sorted examples good in quality? For example, examples with unintended background, or missing body parts should be discarded when doing the sorting. And always use the same, accurate Detector for both example generating and behavioral analysis.

3. The Categorizer you set may be too complex. Try complexity level 2 or 3 for Animation Analyzer and 3 or 4 for Pattern Recognizer. Keep 32 as input shape for Animation Analyzer and try 64 for Pattern Recognizer. And make sure to use the same STD value for both generating behavior examples and training Categorizers. 

4. The mice are mostly white / black / brown / grey. I don't think the color information is important to the behavior classification. So just use Grayscale for the Animation Analyzer. I'm also a little concern that you used STD 0, which may include too many details (including unnecessary ones). Would it possible to show me some pattern images of scratching, body grooming, and hind paw liking? I'm worried that the patten images with STD 0 may be too noisy to distinguish those three behaviors. In this regard, I also suggest you to try 64 as input shape of Animation Analyzer and try 32 for Pattern Recognizer, as this setting puts more weight on Animation Analyzer (that analyzes animations) than Pattern Recognizer (that analyzes pattern images). But since the behavior examples are already sorted, generating and sorting new examples need additional efforts, let's first try other ways to see whether they can solve the problem.

**Reply by rombony** [reply]
We're going to follow your advice and sort our behavior examples to reduce their number, deleting examples that are too close from each other. As we've generated successive examples (of 25 ms every 25 ms) using the entire videos duration, all of our examples follow each other, resulting in a succession of sequences made of very similar examples. We're going to keep only one example in each of these sequences.

We're also going to retrain the categorizer with the parameters you've recommended, but can you give us more information on how to define the STD value? We had tried extreme values during our trials on a much smaller dataset (0, 75 and 255) and did not notice much difference. 

You'll find attached a zip file with some behavior examples. We found out that the size in pixels of the pattern images, and the line width on those patterns, are not always the same. Is it normal? We included examples illustrating this in the zip file. 

About the detector, we used the same one for the generation of examples and the final analysis and encountered no problems with detection, besides occasional background inclusion that we have also included in the zip file. You'll also find examples of frames extracted from our videos before using the detector in case this is of interest to you. Do you think the resolution is too low? Do you have any idea why LabGym's background suppression detection didn't work (paws were missing on most examples)?

[behavior examples.zip](https://github.com/umyelab/LabGym/files/14726996/behavior.examples.zip)

**Reply by yujiahu415** [reply]
Thank you for showing me the behavior examples. As you can see, the pattern images for body grooming, hind paw licking, and scratching are just too similar. This is probably the reason why the Categorizer trained on these pattern images didn't perform well. You mentioned the pattern images with different STD values didn't show differences, this is weird, and is not supposed to happen. Different STD values should show different levels of details within the body outlines. Would it be possible for you to show two pattern images with 0 or 100 STD, for same animal and behavior episode? I need to confirm whether this is a bug. Thank you!

The resolution in animations is good. And the animations show subtle differences among the three behaviors, which can be learned by the Categorizer to distinguish the three behaviors. So for your data, to distinguish the three behaviors, you need to put more weight on animations, rather than pattern images. Ideally you can re-generate behavior examples with STD 50 or 100 (if the STD works properly) and re-sort the examples. But this takes additional effort. So you probably can try train a Categorizer with heavy weight on Animation Analyzer (for analyzing animations) and light weight on Pattern Recognizer (for analyzing pattern image). For example, set AA complexity level 3 or higher, input shape 64 or higher, set PR complexity level 3 or lower, input shape 16 or even 8.

The occasional background inclusion with using Detectors is fine. The reason for background-subtraction-based detection method not working well could be the illumination changes in the video, or the time window used for extracting background is not probably selected. Not sure about this as I haven't seen the videos.

**Reply by rombony** [reply]
Thank you,  we'll try to train the Categorizer by giving more weight to the Animation Analyzer. If that's not sufficient, we'll re-generate our behavior examples by setting higher STD values but we don't know which one would be the best. You'll find a zip file containing behavior examples generated with different STD values (0, 50 and 100), so you can see the difference. 

[STD 0_50_100.zip](https://github.com/umyelab/LabGym/files/14791643/STD.0_50_100.zip)

**Reply by yujiahu415** [reply]
Thank you for showing me the examples! I would suggest to use STD 100 if you want to re-generate the behavior examples. But you can first try to tune the Categorizer settings to see whether it can achieve satisfying performance. And let me know if you have any questions in that process.

**Comment by yujiahu415** [comment]
And also make sure to make the training examples in each behavioral category as diverse as possible--including different scenarios or variants of the behavior and excluding redundant / repeated examples (e.g., if two examples have overlapping frames, just keep one).
